cli/doctest_runner.rs

```diff
@@ -0,0 +1,358 @@
+) -> Result<String, AnyError> {
+  let mut test_file =  "import * as assert from \"https://deno.land/std@0.70.0/testing/asserts.ts\";\n".to_string();
+  let mut import_set = HashSet::new();
+
+  let cwd = std::env::current_dir()?;
+  let cwd_url_str = Url::from_directory_path(cwd)
+    .map(|url| url.to_string())
+    .unwrap_or_else(|_| "".to_string());
+
+  let tests: String = jsdocs
+    .into_iter()
+    .filter_map(|(loc, example)| {
+      let ex_str = clean_string(&example);
+      let test_tag = extract_test_tag(&ex_str);
+      if test_tag == Some("text") {
+        return  None;
+      }
+      let ignore = test_tag == Some("ignore");
+      extract_imports(&ex_str).into_iter().for_each(|import| { import_set.insert(import.to_string()); });
+      let caption = extract_caption(&ex_str);
+      let code_body = extract_code_body(&ex_str);
+      let is_async = has_await(&ex_str);
+      let res = format!(
+        "\nDeno.test({{\n\tname: \"{} - {} (line {})\",\n\tignore: {},\n\t{} fn() {{\n{}\n}}\n}});\n",
+        loc.filename.replace(&cwd_url_str, ""),
+        caption.unwrap_or(""),
+        loc.line,
+        ignore,
+        if is_async { "async"} else {""},
+        code_body
+    );
+    Some(res)
+    })
+    .collect();
+  let imports_str = import_set.into_iter().collect::<Vec<_>>().join("\n");
+  test_file.push_str(&imports_str);
+  test_file.push_str("\n");
+  test_file.push_str(&tests);
+
+  let options = if let Some(filter) = filter {
+    json!({ "failFast": fail_fast, "reportToConsole": !quiet, "disableLog": quiet, "isDoctest": true, "filter": filter })
+  } else {
+    json!({ "failFast": fail_fast, "reportToConsole": !quiet, "disableLog": quiet, "isDoctest": true })
+  };
+
+  let run_tests_cmd = format!(
+    "\n// @ts-ignore\nDeno[Deno.internal].runTests({});\n",
+    options
+  );
+
+  test_file.push_str(&run_tests_cmd);
+  Ok(test_file)
+}
+
+fn clean_string(input: &str) -> String {
+  input
+    .lines()
+    .map(|line| {
+      if line.trim().starts_with('*') {
+        &line.trim()[1..]
+      } else {
+        line.trim()
+      }
+    })
+    .filter(|line| !line.is_empty())
+    .collect::<Vec<_>>()
+    .join("\n")
+}
+
+pub fn is_supported_doctest(path: &Path) -> bool {
+  let valid_ext = ["ts", "tsx", "js", "jsx"];
+  path
+    .extension()
+    .and_then(OsStr::to_str)
+    .map(|ext| valid_ext.contains(&ext) && !is_supported(path))
+    .unwrap_or(false)
+}
+
+fn extract_test_tag(input: &str) -> Option<&str> {
+  TEST_TAG_PATTERN
+    .captures(input)
+    .and_then(|m| m.get(1).map(|c| c.as_str()))
+}
+
+fn extract_caption(input: &str) -> Option<&str> {
+  CAPTION_PATTERN
+    .captures(input)
+    .and_then(|m| m.get(1).map(|c| c.as_str()))
+}
+
+fn extract_imports(input: &str) -> Vec<&str> {
+  IMPORT_PATTERN
+    .captures_iter(input)
+    .filter_map(|caps| caps.get(0).map(|m| m.as_str()))
+    .collect()
+}
+
+fn has_await(input: &str) -> bool {
+  AWAIT_PATTERN.find(input).is_some()
+}
```

Comment:
Those utils seems a little brittle 😬  we should probably leverage SWC for this purpose, but I guess it's alright for first pass

Reply:
Yeah I think so too, will have to parse each of the examples with swc